File: SortTimeComparison/SortingAlgorithms.cs
namespace SortTimeComparison
{
    public static class SortingAlgorithms
    {
        public static void BubbleSort(int[] Array)
        {
            bool swapped = false;

            for (int i = 0; i < Array.Length; i++)
            {
                for (int j = 0; j < Array.Length - i - 1; j++)
                {
                    if (Array[j] > Array[j + 1])
                    {
                        int buf = Array[j];
                        Array[j] = Array[j + 1];
                        Array[j + 1] = buf;
                        swapped = true;
                    }
                }
                if (!swapped)
                {   
                    break;
                }
            }
        }

        public static void InsertionSort(int[] Array)
        {
            for (int i = 1; i < Array.Length; i++)
            {
                int key = Array[i];
                int j = i - 1;

                while (j >= 0 && Array[j] > key)
                {
                    Array[j + 1] = Array[j];
                    j--;
                }

                Array[j + 1] = key;
            }
        }

        public static void QuickSort(int[] Array, int minInd, int maxInd)
        {
            if (minInd < maxInd)
            {
                int pivot = GetPivot(Array, minInd, maxInd);

                QuickSort(Array, minInd, pivot - 1);
                QuickSort(Array, pivot + 1, maxInd);
            }

        }

        private static int GetPivot(int[] Array, int minInd, int maxInd)
        {
            int pivot = Array[maxInd];
            int i = minInd - 1;

            for (int j = minInd; j < maxInd; j++)
            {
                if (Array[j] <= pivot)
                {
                    i++;
                    Swap(Array, i, j);

                }
            }
            Swap(Array, i + 1, maxInd);

            return i + 1;
        }

        private static void Swap(int[] Array, int i, int j)
        {
            int buf = Array[i];
            Array[i] = Array[j];
            Array[j] = buf;
        }
    }
}

File: lab_1_cpp/SortingAlgorithms.cpp
//
// Created by Алоев Артем on 17.02.2025.
//

#include "SortingAlgorithms.h"

#include <vector>

void SortingAlgorithms::BubbleSort(vector<int> &array) {
    int n = array.size();
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (array[j] > array[j + 1]) {
                int buf = array[j];
                array[j] = array[j + 1];
                array[j + 1] = buf;
            }
        }
    }
}

void SortingAlgorithms::InsertionSort(vector<int> &array) {
    int n = array.size();
    for (int i = 1; i < n; i++) {
        int key = array[i];
        int j = i - 1;

        while (j >= 0 && array[j] > key) {
            array[j + 1] = array[j];
            j--;
        }

        array[j + 1] = key;
    }
}

void SortingAlgorithms::SelectionSort(vector<int> &array) {
    int n = array.size();
    for (int i = 0; i < n - 1; i++) {

        int min_idx = i;
        for (int j = i + 1; j < n; j++) {
            if (array[j] < array[min_idx]) {
                min_idx = j;
            }
        }

        if (min_idx != i) {
            int buf = array[i];
            array[i] = array[min_idx];
            array[min_idx] = buf;
        }
    }
}